### PR TITLE
[Sapling] Incremental merkle tree backport.

### PR DESCRIPTION
**Pull request by furszy**
This is coming from #1798, focused on include the incremental merkle tree primitive with all of its unit tests.
This base primitive class is almost one to one with upstream.

Decoupled the following commits:

* Incremental Merkle Tree + merkletree tests back port —> 762d5f10
* Rename sapling util file to not shadow src/util file --> cae89d3b

### CODE REVIEW

**Comment by furszy** [comment]
This is fixing #1798 travis issues. took a while to decouple it but it's is ready for review.

File: src/sapling/incrementalmerkletree.cpp [hunk 1]
```diff
@@ -0,0 +1,1155 @@
+// Copyright (c) 2016-2018 The Zcash developers
+// Copyright (c) 2020 The PIVX developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
```
Comment (Fuzzbawls) [inline src/sapling/incrementalmerkletree.cpp:9]: `#include optional.h` to have our own wrapper available for calls below

File: src/sapling/incrementalmerkletree.cpp [hunk 2]
```diff
@@ -0,0 +1,1155 @@
...
+    if (!left) {
+        // Set the left leaf
+        left = obj;
+    } else if (!right) {
+        // Set the right leaf
+        right = obj;
+    } else {
+        // Combine the leaves and propagate it up the tree
+        boost::optional<Hash> combined = Hash::combine(*left, *right, 0);
```
Comment (Fuzzbawls) [inline src/sapling/incrementalmerkletree.cpp:937]: `Optional<Hash>...`

File: src/sapling/incrementalmerkletree.cpp [hunk 3]
```diff
@@ -0,0 +1,1155 @@
...
+bool IncrementalMerkleTree<Depth, Hash>::is_complete(size_t depth) const {
+    if (!left || !right) {
+        return false;
+    }
+
+    if (parents.size() != (depth - 1)) {
+        return false;
+    }
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
```
[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply src/sapling/incrementalmerkletree.cpp:973]: `Optional<Hash>&`

File: src/sapling/incrementalmerkletree.cpp [hunk 4]
```diff
@@ -0,0 +1,1155 @@
...
+        if (skip) {
+            skip--;
+        } else {
+            return 0;
+        }
+    }
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
```
[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply src/sapling/incrementalmerkletree.cpp:1004]: Optional<Hash>&

File: src/sapling/incrementalmerkletree.cpp [hunk 5]
```diff
@@ -0,0 +1,1155 @@
...
+    PathFiller<Depth, Hash> filler(filler_hashes);
+
+    Hash combine_left =  left  ? *left  : filler.next(0);
+    Hash combine_right = right ? *right : filler.next(0);
+
+    Hash root = Hash::combine(combine_left, combine_right, 0);
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
```
[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply src/sapling/incrementalmerkletree.cpp:1032]: Optional<Hash>&

File: src/sapling/incrementalmerkletree.cpp [hunk 6]
```diff
@@ -0,0 +1,1155 @@
...
+        index.push_back(true);
+        path.push_back(*left);
+    } else {
+        index.push_back(false);
+        path.push_back(filler.next(0));
+    }
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
```
[the remark this answers is not in the document]
Comment (Fuzzbawls) [reply src/sapling/incrementalmerkletree.cpp:1075]: Optional<Hash>&

**Review by Fuzzbawls — CHANGES_REQUESTED** [review]
Could use some cleanup regarding our own standards and helpers/wrappers so as to not re-introduce new boost dependencies.

**Comment by furszy** [comment]
Done, styling updated to our sources conventions 👍 .

**Review by random-zebra — APPROVED** [review]
ACK 626b0d93a3eef3abd3f481779ea507919b3eacb5

**Review by Fuzzbawls — APPROVED** [review]
utACK 626b0d93a3eef3abd3f481779ea507919b3eacb5

**Pull request merged**